docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,345 @@
+
+// Storing accessor somewhere...
+let tenant_id = store_my_tenant_pls!(accessor);
+```
+
+Tenant cancellation:
+
+```rust
+struct Tenant {
+    gc_control: LoopedTaskController,
+    compaction_control: LoopedTaskController,
+    drop_sender: DropSender,
+    // ...
+}
+
+impl Drop for Tenant {
+    // Note: no task_mgr!
+    fn drop(&mut self) {
+        // ...
+        BACKGROUND_RUNTIME.spawn(move || {
+            gc_control.break();
+            compaction_control.break();
+
+            gc_control.wait().await;
+            compaction_control.wait().await;
+
+            drop(drop_sender);
+            info("tenant drop finished")
+        })
+    }
+}
+```
+
+Working with a tenant. GC example:
+
+```rust
+'iteration loop {
+    trace!("waking up");
+    with_tenant(tenant_id, |tenant| {
+        // Assuming async closures are present
+        // Note the sleep outside the closure!
+        let sleep_duration = tenant_mgr::with_tenant_async(tenant_id, async move |tenant| {
+            let gc_period = tenant.get_gc_period();
+            let gc_horizon = tenant.get_gc_horizon();
+            let mut sleep_duration = gc_period;
+            if gc_horizon > 0 {
+                if let Err(e) =
+                    tenant.gc_iteration(None, gc_horizon, tenant.get_pitr_interval(), false).await
+                {
+                    sleep_duration = wait_duration;
+                    error!("Gc failed, retrying in {:?}: {e:#}", sleep_duration);
+                }
+            }
+            sleep_duration
+        })
+        .await?;
+
+        let waiter = DurationWaiter::new(sleep_duration);
+        'sleep loop {
+            tokio::select! {
+                state = self.control_flow.get_state() => {
+                    match state {
+                        LoopedTaskControlFlow::Continue => continue 'sleep,
+                        LoopedTaskControlFlow::Pause => 'pause loop {
+                            match self.control_flow.get_state() {
+                                LoopedTaskControlFlow::Continue => continue 'sleep,
+                                // The thing is it's possible to have `Pause` state here
+                                // again, because we were fast enough to have transitions
+                                // Pause -> Continue -> Pause. But, it could be fixed by
+                                // the third loop if needed, though it's better to write
+                                // operator code in a such way we won't get this state twice.
+                                LoopedTaskControlFlow::Pause => continue 'pause,
+                                LoopedTaskControlFlow::Break => break 'iteration,
+                            }
+                        }
+                        LoopedTaskControlFlow::Break => break 'iteration,
+                    }
+                },
+                _ = waiter.wait() => break 'sleep,
+            }
+        }
+    })
+}
+```
```

Comment:
Adding my "request changes" comment as a thread, so it will be easier to respond.

For many times I've tried to understand this RFC and I will try to continue to read and understand proposal, but in general I must be opposed to adding this much new infrastructure and am worried that we will gain more hidable bugs. New infrastructure such as `LoopedTaskController`, `LoopedTaskControlFlow`, `DurationWaiter`.

As I already mentioned, `tokio::task::JoinSet` and `tokio_util::sync::CancellationToken` (hierarchy) should handle many of these issues (latter also works in sync contexts, with caveats, but currently we have no cancelling for those, but [the code looks like it's cancellable](https://github.com/neondatabase/neon/blob/1c8e0804890d1b0d1107240be92a8f6c30233e91/pageserver/src/tenant/timeline.rs#L995-L1027)). What is currently forbidding running `spawn_blocking` operations, such as the gc or compaction is that we cannot move `shutdown_watcher` to tasks (easily). This does have a trivial solution, but fixing that would be suboptimal.

Instead of using RAII constructs what might work to simplify is to have an actor based model where each tenant would be an actor which owns all of it's state (I'd like to prototype this sort of approach) so it is able to react to failures but also requests which would have to cancel any right running operation. Mutual exclusion of different tasks would not have to happen, as in, we could simply not have two mutually exclusive things at the same time. This would also allow us to have a more "single threaded" implementation or `&mut` operations, which would remove many of the problems I think this RFC aims to solve.